### DISCUSSION TITLE
Unable To Link Alexa Skill

### DISCUSSION THREAD

**Discussion by alienstar-uk** [koush/scrypted#1299] [Q&A]
Scrypted Server is installed and working and connection to the Scrptted Cloud and Hikvision cameras. Whenever I try to enable the Amazon Alexa skill it fails with the message 

"
Unable to link the skill at this time
Please try again later.
"

As I use a UK Amazon account and Alexa devices I use the following URL to enable the skill:

https://www.amazon.co.uk/ClockworkMod-Scrypted-Home-Automation/dp/B09WHHD9VN/ref=sr_1_1?keywords=scrypted&qid=1706554229&s=digital-skills&sr=1-1

After a brief period of trying to link the skill it fails with the "Unable to link the skill at this time" error message. 

In the console (with debug enabled on the Alexa plugin) it displays the following:

"
########################
2/5/2024, 10:21:50 AM
########################
received directive from alexa Alexa.Authorization/AcceptGrant {
  directive: {
    header: {
      namespace: 'Alexa.Authorization',
      name: 'AcceptGrant',
      messageId: '13439667-3d29-4426-b3c9-46e801e9c4f2',
      payloadVersion: '3'
    },
    payload: { grant: [Object], grantee: [Object] }
  }
}
response to alexa directive 200 {
  event: {
    header: {
      namespace: 'Alexa.Authorization',
      name: 'AcceptGrant.Response',
      messageId: 'bbc56acb-f031-453d-b7d4-41d2a0c76e93',
      payloadVersion: '3'
    },
    payload: {}
  }
}
"

I have attempted enabling the skill multiple times over the past 5 days from an Edge browser on Windows 10 and on IOS 17.3, all wih the same failure. 

Desktop: 
 - Windows 10
 - Edge Browser Version 121.0.2277.98 (Official build) (64-bit)

Smartphone
 - Device: iPhone 15 Pro Max
 - OS: IOS 17.3
 - Edge Browser  Version 121.2277.99

Scrypted Version Version 0.91.6 running on QNAP Docker Debian Image

**Comment by koush** [comment]
try using the mobile app

**Reply by alienstar-uk** [reply]
Tried the mobile app but this also fails.

**Reply by alienstar-uk** [reply]
I recreated the QNAP container and changed the container network mode to be Custom - Host (as per your QNAP Container install instructions) rather than a virtual IP address (which I had used to prevent port clashes) and this worked.

**Comment by blairscarlett** [comment]
I am having a similiar problem. I have tried from both the mobile app and several computers. The only difference architecture difference is that Scrypted is running on Ubuntu within HyperV. I have confirmed that Scrypted Cloud port forwarding is working. I had this working on a previous HyperV yet have to rebuild it and cannot get the Alexa skill to link.

**Reply by foxy82** [reply]
I'm also suffering the same thing - using Proxmox LXC container - scrypted is saying the port forward is working and checking locally in my network I can see the container has the port open.

I see this in the cloud log just before the Alexa App/Alexa Website says linking failed: 
```50.219.188.186 /web/component/alexa/endpoint```

The Alexa plugin doesn't add anything during linking but it does show this :

```
plugin loaded
startup failed Error: 'tokenInfo' is undefined
    at h.getAccessToken (/src/main.ts:418:19)
    at h.this [as syncEndpoints] (/src/main.ts:332:40)
    at h.start (/src/main.ts:99:17)
```

**Reply by skynet01** [reply]
I've been having the same issue for month now, it used to work ok for me. You did get further than I did by actually getting an error shown. Looks like a bug, maybe @koush can trouble shoot here? I am on amazon US server and tried desktop and app version and get same "please try again" error

**Reply by blairscarlett** [reply]
I just tried linking the skill and it worked today. I didnt change anything on my config so I have to assume it was Amazon. Also worthwhil eto mention, I use amazon.ca  Glad it works now!

**Reply by skynet01** [reply]
Still the same issue for me on the US site. Port forwarding works and I can access my instance from the cloud. 
![image](https://github.com/koush/scrypted/assets/8117544/6c908836-16f4-416b-8e4c-3be9850fb0f7)

**Reply by foxy82** [reply]
I downgraded to an older version that worked for me in the past. I don't see the token error now but in the Alexa plugin console logs messages I see the Alexa messages arrive after Amazon has given up trying to link the skill - so I wonder if it is just too slow / too much load on the scrypted cloud plugin.

**Comment by skynet01** [comment]
What version did you downgrade to? Mine still has an error 

```
unhandledRejection Error: 'tokenInfo' is undefined
    at h.getAccessToken (/src/main.ts:418:19)
    at h.addAccessToken (/src/main.ts:213:40)
    at h.deviceListen (/src/main.ts:207:27)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
```

**Comment by InnovativeDevelopers** [comment]
Same issue here.  Has anyone in the UK managed to enable this?

**Comment by InnovativeDevelopers** [comment]
Fixed it finally!  There's an ongoing issue with docker on Synology where networking starts to run slow.  I manually set the DNS to Cloudflare (anything will do) and everything started to work.  Weird!

**Reply by alienstar-uk** [reply]
I am in the UK and have got it to work using a QNAP container. After initial problems I recreated the QNAP container and changed the container network mode to be Custom - Host (as per the QNAP Container install instructions) rather than a virtual IP address (which I had used to prevent port clashes) and this worked. Have been running it like this since Feb 18th and have had no problems. The camera feed is only about 2 seconds behind the live feed - which is far better than the MONOCLE solution I used previously where it could be up to 10 seconds behind and frequently suffered from "buffering" messages.

**Comment by davidecavestro** [comment]
Still unable to link... I've tried both from desktop and from mobile, using a host-networked container on linux - with port forwarded from router - to enable the skill on _amazon.it_.

The plugin partially gets its configuration (i.e. endpoint _api.eu.amazonalexa.com_), but seems to fail acquiring the token.

Aoart from the error message from the amazon web page and the 400 response on the browser, the scrypted logs show

```
received directive from alexa Alexa.Authorization/AcceptGrant {
  directive: {
    header: {
      namespace: 'Alexa.Authorization',
      name: 'AcceptGrant',
      messageId: '...',
      payloadVersion: '3'
    },
    payload: { grant: [Object], grantee: [Object] }
  }
}
scrypted  |
response to alexa directive 200 {
  event: {
    header: {
      namespace: 'Alexa.Authorization',
      name: 'AcceptGrant.Response',
      messageId: '...',
      payloadVersion: '3'
    },
    payload: {}
  }
}
```

UPDATE: as already noted at https://github.com/koush/scrypted/discussions/1396#discussioncomment-9371762 the logged messages for _received directive from alexa_ and _response to alexa directive 200_ are printed to the logs slightly after Amazon has given up linking the skill

UPDATE2: now I cannot reproduce the issue as after I restarted from scratch, I always get a _Server Unreachable - Your server is not reachable by Scrypted Cloud_ error message when trying to login through the cloud plugin even if the port forward is working properly and I can access the UI using that port from the web. It's weird

**Reply by davidecavestro** [reply]
Got it: I had an obsolete, unreachable DNS entry in _/etc/resolv.conf_ on the host, slowing down name resolution on the container.

Once removed, the linking process completed successfully.
A clear symptom is `curl https://manage.scrypted.app` from the container: it should return immediately the html. If not, there's a slowdown.

**Comment by skynet01** [comment]
Yeah same thing here still.

**Reply by davidecavestro** [reply]
I'm just wondering if we should [post to discord](https://discord.com/channels/882329362295316500/955941045907095562) instead

**Comment by davidecavestro** [comment]
IMHO When the message _received directive from alexa_ and its response counterpart are shown AFTER the Alexa skill gave up with error, it's probably a timing issue.

For me it was an unreachable entry in _/etc/resolv.conf_, propagated to the container, so that the name resolution was really slow.
Once removed it worked.

In fact with the unreachable DNS even issuing a command like `curl https://manage.scrypted.app` from the container took some seconds

**Comment by skynet01** [comment]
i can confirm that this solved the issue. Just edited my /etc/resolv.conf in the scrypted docker container to not have one of my network dns servers and the authentication worked
